### PR TITLE
Bls12 377 gadgets for Groth16

### PR DESCRIPTION
**Pull request by dtebbs**

### CODE REVIEW

File: libzecale/circuits/pairing/bls12_377_pairing.hpp [hunk 1]
```diff
@@ -6,33 +6,52 @@
...
+public:
+    using FieldT = libff::Fr<ppT>;
+
+    std::shared_ptr<libsnark::pb_linear_combination<FieldT>> _Px;
+    std::shared_ptr<libsnark::pb_linear_combination<FieldT>> _Py;
+
+    bls12_377_G1_precomputation();
+
+    // Construct a populated G1_precomputation from a value. All terms are
+    // created as constants, generating no cost in the circuit.
```
Comment (AntoineRondelet) [inline libzecale/circuits/pairing/bls12_377_pairing.hpp:31]: `generating no cost in the circuit`, what do you mean here? Do you mean "this function does not add any gates in the circuit", hence "doesn't alter the prover perfs"?

File: libzecale/circuits/pairing/bls12_377_pairing.hpp [hunk 2]
```diff
@@ -441,6 +464,71 @@ class bls12_377_final_exp_last_part_gadget
...
+// Wrapper around final_exp gadgets with interface expected by the groth16
+// gadgets. `result_is_one` is constrained to a boolean (0 or 1), and set in
+// 'generate_r1cs_witness' based on the output value of the final
+// exponentiation (if final exp == 1, `result_is_one` is set to 1, otherwise
+// `result_is_one` is set to 0).
+//
+// Note that the constraints on the final exp output are ONLY enforced when
+// `result_is_one` == 1. In otherwords, it is infeasible to generate valid
+// inputs such that the final exp output is not equal to 1 and result_is_one ==
+// 1. However, it IS possible to generate inputs such that final_exp == 1 but
```
Comment (AntoineRondelet) [inline libzecale/circuits/pairing/bls12_377_pairing.hpp:476]: Shouldn't it be "However, it IS possible to generate inputs such that the final exp output is not equal to 1 but `result_is_one` == 0"?

File: libzecale/circuits/pairing/bls12_377_pairing.tcc [hunk 3]
```diff
@@ -105,420 +194,484 @@ template<typename ppT>
...
-    , out_R(pb, " R")
+    , _in_R(R)
+    , _out_R(out_R)
+    , _out_coeffs(coeffs)
 
     // A = Rx * Ry / 2
-    , A(pb, FMT(annotation_prefix, " A"))
-    , check_A(
-          pb, in_R.X, in_R.Y, A * FqT(2), FMT(annotation_prefix, " check_A"))
+    , _A(pb,
```
Comment (AntoineRondelet) [inline libzecale/circuits/pairing/bls12_377_pairing.tcc:206]: Let's be consistent on how we build these gadgets here. I most parts of the code base, smart ptrs are used and passed as args before being `reset/new`'ed in the parent gadget's constructor. This contrasts with the existing code.

File: libzecale/circuits/pairing/bls12_377_pairing.tcc [hunk 4]
```diff
@@ -105,420 +194,484 @@ template<typename ppT>
...
+         _in_R.X,
+         _in_R.Y * FqT(2).inverse(),
+         Fqe_variable<ppT>(pb, FMT(annotation_prefix, " Rx*Ry/2")),
+         FMT(annotation_prefix, " _A"))
 
     // B = Ry^2
-    , B(pb, FMT(annotation_prefix, " B"))
-    , check_B(pb, in_R.Y, B, FMT(annotation_prefix, " check_B"))
+    , _B(pb,
```
Comment (AntoineRondelet) [inline libzecale/circuits/pairing/bls12_377_pairing.tcc:213]: Another place where https://github.com/clearmatics/zecale/issues/20 would apply.

 - `_B` ->`compute_B` (same for the annotation)
 - `Ry^2` (the output wire annotation) -> `B`

Same applies for the other gadgets in this file

File: libzecale/serialization/proto_utils.hpp [hunk 5]
```diff
@@ -11,12 +11,13 @@
 namespace libzecale
 {
 
-template<typename ppT, typename snarkT, typename snarkHandlerT>
-transaction_to_aggregate<ppT, snarkT> transaction_to_aggregate_from_proto(
+template<typename ppT, typename apiHandlerT>
+transaction_to_aggregate<ppT, typename apiHandlerT::snarkT>
```
Comment (AntoineRondelet) [inline libzecale/serialization/proto_utils.hpp:15]: Why do we have `apiHandler::snarkT` and `wVerifierT::SnarkT` (i.e. why do we have "small and big" `S` for snark)?

File: libzecale/tests/circuits/bls12_377_pairing_test.cpp [hunk 6]
```diff
@@ -284,23 +278,53 @@ TEST(BLS12_377_PairingTest, PrecomputeGadgetTest)
...
+        libff::bls12_377_ate_precompute_G2(Q);
+
+    // Circuit with precompute gadget
+    libsnark::protoboard<libff::Fr<wpp>> pb;
+    libsnark::G2_variable<wpp> Q_var(pb, "Q");
+    libzecale::bls12_377_G2_precomputation<wpp> Q_prec_var;
+    const size_t num_primary_inputs = pb.num_inputs();
+    pb.set_input_sizes(num_primary_inputs);
+    libzecale::bls12_377_G2_precompute_gadget<wpp> precompute_gadget(
+        pb, Q_var, Q_prec_var, "bls12_317 Q2 precompute gadget");
```
Comment (AntoineRondelet) [inline libzecale/tests/circuits/bls12_377_pairing_test.cpp:316]: `precompute_gadget`

File: libzecale/tests/circuits/miller_loop_gadgets_test.cpp [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+    test_mnt_e_times_e_times_e_over_e_miller_loop<libff::mnt4_pp>(
+        " test_eee_over_e_miller_loop_mnt4");
+    test_mnt_e_times_e_times_e_over_e_miller_loop<libff::mnt6_pp>(
+        " test_eee_over_e_miller_loop_mnt6");
+}
+
+TEST(MillerLoopGadgets, TestBlsEEEoverEmillerLoop)
+{
+    using opp = libff::bw6_761_pp;   // outer pairing
+    using ipp = libff::bls12_377_pp; // inner pairing
```
Comment (AntoineRondelet) [inline libzecale/tests/circuits/miller_loop_gadgets_test.cpp:214]: `nested`/`wrapping`

File: libzecale/circuits/pairing/bls12_377_pairing.hpp (hunk 2 above)
Comment (dtebbs) [reply libzecale/circuits/pairing/bls12_377_pairing.hpp:476]: If `result_is_one == 0`, all checks are effectively disabled and so it IS possible that `final_exp == 1`.
I.e. (if I've understood) `result_is_one == 1` means that the final exp value is 1, but `result_is_one == 0` does not say anything about the value of final exp.

File: libzecale/circuits/pairing/bls12_377_pairing.tcc (hunk 4 above)
Comment (dtebbs) [reply libzecale/circuits/pairing/bls12_377_pairing.tcc:213]: This should be covered in a later change.

File: libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc [hunk 8]
```diff
@@ -314,28 +314,28 @@ r1cs_gg_ppzksnark_verifier_process_vk_gadget<ppT>::
...
 
-    pvk.vk_generator_g2_precomp.reset(new libsnark::G2_precomputation<ppT>());
-    pvk.vk_beta_g2_precomp.reset(new libsnark::G2_precomputation<ppT>());
-    pvk.vk_delta_g2_precomp.reset(new libsnark::G2_precomputation<ppT>());
+    pvk.vk_generator_g2_precomp.reset(new G2_precomputation<ppT>());
+    pvk.vk_beta_g2_precomp.reset(new G2_precomputation<ppT>());
+    pvk.vk_delta_g2_precomp.reset(new G2_precomputation<ppT>());
 
-    compute_vk_alpha_g1_precomp.reset(new libsnark::precompute_G1_gadget<ppT>(
+    compute_vk_alpha_g1_precomp.reset(new G1_precompute_gadget<ppT>(
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc:323]: This is not quite renaming.  The original code is referring to a concrete gadget, but we now need to select it based on `ppT`.  The libsnark version of pairing_params does not contain an alias for this, so I've had to add it to the zecale `pariring_params.hpp` alongside all the other aliases we have.  So here, this is switching from a hard-coded specific gadget to the parameterized alias.

[the remark this answers is not in the document]
Comment (AntoineRondelet) [reply libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc:323]: Oh right, indeed! Got you

File: libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc [hunk 9]
```diff
@@ -314,28 +314,28 @@ r1cs_gg_ppzksnark_verifier_process_vk_gadget<ppT>::
...
         *pvk.vk_alpha_g1_precomp,
         FMT(annotation_prefix, " compute_vk_alpha_g1_precomp")));
 
-    pvk.vk_generator_g2_precomp.reset(new libsnark::G2_precomputation<ppT>(
+    pvk.vk_generator_g2_precomp.reset(new G2_precomputation<ppT>(
         pb,
         libff::G2<other_curve<ppT>>::one(),
         FMT(annotation_prefix, " vk_generator_g2_precomp")));
-    compute_vk_beta_g2_precomp.reset(new libsnark::precompute_G2_gadget<ppT>(
+    compute_vk_beta_g2_precomp.reset(new G2_precompute_gadget<ppT>(
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc:333]: See above - there is no alias to select precompute gadgets in libsnark so in order to make this code generic we need to use the new aliases in libzecale.

File: libzecale/circuits/pairing/bls12_377_pairing.hpp (hunk 1 above)
Comment (dtebbs) [reply libzecale/circuits/pairing/bls12_377_pairing.hpp]: reworded the comment

File: libzecale/tests/circuits/miller_loop_gadgets_test.cpp (hunk 7 above)
Comment (dtebbs) [reply libzecale/tests/circuits/miller_loop_gadgets_test.cpp]: renamed

File: libzecale/circuits/pairing/bls12_377_pairing.tcc (hunk 3 above)
Comment (AntoineRondelet) [reply libzecale/circuits/pairing/bls12_377_pairing.tcc:206]: See: https://github.com/clearmatics/zecale/issues/23

File: libzecale/circuits/aggregator.tcc [hunk 10]
```diff
@@ -49,11 +41,23 @@ namespace libzecale
...
+    typename nppT,
+    typename wppT,
+    typename nSnarkT,
+    typename wVerifierT,
+    size_t NumProofs>
 class aggregator_gadget : libsnark::gadget<libff::Fr<wppT>>
 {
 private:
-    std::array<std::shared_ptr<VerifierGadgetT<wppT>>, NumProofs> verifiers;
+    using wsnarkT = typename wVerifierT::SnarkT;
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/circuits/aggregator.tcc]: clearmatics/zeth#233 addresses some remaining inconsistencies in zeth
updated this PR to address the zecale naming.

File: libzecale/serialization/proto_utils.hpp (hunk 5 above)
Comment (dtebbs) [reply libzecale/serialization/proto_utils.hpp]: addressed above.  all types use lower-case, consistent with classes etc.  `T` for type parameters only.

File: libzecale/tests/circuits/bls12_377_pairing_test.cpp (hunk 6 above)
Comment (dtebbs) [reply libzecale/tests/circuits/bls12_377_pairing_test.cpp]: changed annotation

File: libzecale/circuits/aggregator.tcc (hunk 10 above)
[the remark this answers is not in the document]
Comment (AntoineRondelet) [reply libzecale/circuits/aggregator.tcc]: Thanks

File: libzecale/circuits/pairing/bls12_377_pairing.hpp (hunk 2 above)
Comment (AntoineRondelet) [reply libzecale/circuits/pairing/bls12_377_pairing.hpp:476]: As discussed, seems like we are saying the same thing. If `result_is_one = 0` then the following checks are rendered dummy, which means that the result isn't constrained to a given value.

**Pull request merged**